### ISSUE TITLE
Build/Research Costs and Times incorrect

### ISSUE THREAD

**Issue by dbenage-cx** [freeorion/freeorion#1261]
Build from source: d5f7f74

### Description
Costs and times for buildings and techs are incorrect, seemingly when the time or cost are defined using certain valuerefs.
Building example: Starlane bore - 999k PP / 999 turns
Research example: Planetary Cloud Cover - 999k RP / 5 turns

`git bisect` resulted in 8ac9609
reverting the portion for `universe/Building.cpp` restores the building costs to normal values.

**Comment by geoffthemedio** [comment]
Yeah, that change looks wrong to me.
````
-        if (!source && !m_production_cost->SourceInvariant())
+        if (!source || m_production_cost->SourceInvariant())
             return 999999.9f;
````
The point was: if there is no source and the cost depends on the source, return a default value.

Now it's been changed to: if there is no source or the cost doesn't depend on the source, return a default value. That means that regardless of whether there is a source, if the cost doesn't depend on the source the default value will be used, which is wrong in cases like a constant cost that doesn't depend on the source.

If there is no source but the cost doesn't depend on the source, then things are fine to evaluate the valueref using a null source, since the cost doesn't depend on the source.

There could potentially be a problem is SourceInvariant() isn't implemented correctly, and an attempt is made to access the source within a supposedly-source-invariant valueref, but just not having a source is not a reason to never attempt to evaluate a valueref... many work fine without a valid source object... those are supposed to be the source-invariant ones. (Or at least the source-invariant ones are supposed to not change value depending on the source... there could be some that still attempt to access the source object even if their value doesn't depend on the source object, but that's a bug to fix within them, not here...)

**Comment by LGM-Doyle** [comment]
Sorry.  I misunderstood that.  Your explanation clarifies it.

I thought that the intent was to return a reasonable default value when the source was missing or the whole thing was source invariant.  I see that this really is an error condition.

I will provide a solution.

**Comment by Vezzra** [comment]
@dbenage-cx, a fix (PR #1262) has been merged, can you check if that solves the issue for you?

**Comment by dbenage-cx** [comment]
@LGM-Doyle Thanks for providing a fix
